src/sn_handover.rs

```diff
@@ -0,0 +1,174 @@
+    pub gen: UniqueSectionId,
+}
+
+impl<T: Proposition> Handover<T> {
+    pub fn from(secret_key: SecretKeyShare, elders: BTreeSet<PublicKeyShare>, gen: UniqueSectionId) -> Self {
+        Handover::<T> {
+            consensus: Consensus::<T>::from(secret_key, elders),
+            gen,
+        }
+    }
+
+    pub fn random(rng: impl Rng + CryptoRng, gen: UniqueSectionId) -> Self {
+        Handover::<T> {
+            consensus: Consensus::<T>::random(rng),
+            gen,
+        }
+    }
+
+    pub fn propose(&mut self, proposal: T) -> Result<SignedVote<T>> {
+        let vote = Vote {
+            gen: self.gen,
+            ballot: Ballot::Propose(proposal),
+        };
+        let signed_vote = self.sign_vote(vote)?;
+        self.validate_signed_vote(&signed_vote)?;
+        Ok(self.cast_vote(signed_vote))
+    }
+
+    // Get someone up to speed on our view of the current votes
+    pub fn anti_entropy(&self) -> Vec<SignedVote<T>> {
+        info!(
+            "[HDVR] anti-entropy from {:?}",
+            self.public_key_share()
+        );
+
+        self.consensus.votes.values().cloned().collect()
+    }
+
+    fn resolve_votes(&self, votes: &BTreeSet<SignedVote<T>>) -> Option<T> {
+        let (winning_proposals, _) = self
+            .count_votes(votes)
+            .into_iter()
+            .max_by_key(|(_, count)| *count)
+            .unwrap_or_default();
+
+        // we need to choose one deterministically
+        // proposals are comparable because they impl Ord so we arbitrarily pick the max
+        winning_proposals.into_iter().max()
+    }
+
+    pub fn public_key_share(&self) -> PublicKeyShare {
+        self.consensus.public_key_share()
+    }
+
+    pub fn handle_signed_vote(
+        &mut self,
+        signed_vote: SignedVote<T>,
+    ) -> Result<Option<SignedVote<T>>> {
+        self.validate_signed_vote(&signed_vote)?;
+
+        let vote_response = self
+            .consensus
+            .handle_signed_vote(signed_vote, self.gen)?;
+
+        match vote_response {
+            VoteResponse::Broadcast(vote) => {
+                Ok(Some(vote))
+            }
+            VoteResponse::Decided(_vote) => {
+                Ok(None)
+            }
+            VoteResponse::WaitingForMoreVotes => Ok(None),
+        }
+    }
+
+    pub fn sign_vote(&self, vote: Vote<T>) -> Result<SignedVote<T>> {
+        self.consensus.sign_vote(vote)
+    }
+
+    pub fn cast_vote(&mut self, signed_vote: SignedVote<T>) -> SignedVote<T> {
+        self.log_signed_vote(&signed_vote);
+        signed_vote
+    }
+
+    fn log_signed_vote(&mut self, signed_vote: &SignedVote<T>) {
+        self.consensus.log_signed_vote(signed_vote);
+    }
+
+    pub fn count_votes(
+        &self,
+        votes: &BTreeSet<SignedVote<T>>,
+    ) -> BTreeMap<BTreeSet<T>, usize> {
+        self.consensus.count_votes(votes)
+    }
+
+    pub fn validate_signed_vote(&self, signed_vote: &SignedVote<T>) -> Result<()> {
```

Comment:
Are these validations identical to the sn_membership logic?

Reply:
Except for the generation check and the proposal check yes.
```rust
        if vote.gen != self.gen {                                                   // here
            return Err(Error::VoteWithInvalidUniqueSectionId                        // here
                vote_gen: vote.gen,
                gen: self.gen,
            });
        }

        match &vote.ballot {
            Ballot::Propose(proposal) => self.validate_proposal(proposal.clone()),  // and here
 ```

Reply:
I wanted to move them out to `consensus` at first because they see the underlying ballot stuff but the recursive call to `validate_signed_vote` makes it hard, so I left them here for now.

Reply:
what if we front-load all proposal and generation checks before calling `Consensus::handle_signed_vote()`?

I.e. unpack all proposals in a vote with `SignedVote::proposals()` and iterate through each of those validating them before sending the vote down to consensus.

Similarly for generations, check all nested votes have the same generation before forwarding the vote

Reply:
I did the first generation check outside as the check is different for Membership and Handover, but kept the generation checks for nested votes as they are the same and it would require to iterate over Votes twice and I'd prefer to hide the Ballot matching logic from non-consensus code as much as possible.